// @polymer/paper-scroll-header-panel/paper-scroll-header-panel.js
define(["exports","../polymer/polymer-legacy.js","../iron-resizable-behavior/iron-resizable-behavior.js","../polymer/lib/legacy/polymer-fn.js","../polymer/lib/legacy/polymer.dom.js","../polymer/lib/utils/html-tag.js"],function(_exports,_polymerLegacy,_ironResizableBehavior,_polymerFn,_polymerDom,_htmlTag){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.PaperScrollHeaderPanel=void 0;const PaperScrollHeaderPanel=(0,_polymerFn.Polymer)({_template:_htmlTag.html`
    <style>
      :host {
        display: block;
        position: relative;
        overflow: hidden;
      }

      #mainContainer {
        position: absolute;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        -moz-box-sizing: border-box;
        box-sizing: border-box;
        -webkit-overflow-scrolling: touch;
        overflow-x: hidden;
        overflow-y: auto;
        @apply --paper-scroll-header-container; /* deprecated due to incorrect/confusing naming */
        @apply --paper-scroll-header-panel-container;
      }

      #headerContainer {
        position: absolute;
        top: 0;
        right: 0;
        left: 0;
        @apply --paper-scroll-header-panel-header-container;
      }

      .bg-container {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
        overflow: hidden;
      }

      #headerBg {
        @apply --paper-scroll-header-panel-full-header;
      }

      #condensedHeaderBg {
        @apply --paper-scroll-header-panel-condensed-header;
      }

      #headerBg, #condensedHeaderBg {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
        background-repeat: no-repeat;
        background-size: cover;
        background-position: center center;
      }

      #condensedHeaderBg {
        opacity: 0;
      }
    </style>

    <div id="mainContainer">
      <slot id="mainSlot" name="content"></slot>
    </div>
    <div id="headerContainer">
      <div class="bg-container">
        <div id="condensedHeaderBg"></div>
        <div id="headerBg"></div>
      </div>
      <slot id="headerSlot" name="header"></slot>
    </div>
`,is:"paper-scroll-header-panel",behaviors:[_ironResizableBehavior.IronResizableBehavior],properties:{condenses:{type:Boolean,value:!1},noDissolve:{type:Boolean,value:!1},noReveal:{type:Boolean,value:!1},fixed:{type:Boolean,value:!1},keepCondensedHeader:{type:Boolean,value:!1},headerHeight:{type:Number},condensedHeaderHeight:{type:Number,value:0},scrollAwayTopbar:{type:Boolean,value:!1},headerState:{type:Number,readOnly:!0,notify:!0,value:0},_defaultCondsensedHeaderHeight:{type:Number,value:0}},observers:["_setup(headerHeight, condensedHeaderHeight, fixed)","_condensedHeaderHeightChanged(condensedHeaderHeight)","_headerHeightChanged(headerHeight, condensedHeaderHeight)","_condensesChanged(condenses, headerHeight)"],listeners:{"iron-resize":"measureHeaderHeight"},ready:function(){this._scrollHandler=this._scroll.bind(this);this.scroller.addEventListener("scroll",this._scrollHandler);console.warn(this.is,"is deprecated. Please use app-layout instead!")},attached:function(){requestAnimationFrame(this.measureHeaderHeight.bind(this))},get header(){return(0,_polymerDom.dom)(this.$.headerSlot).getDistributedNodes()[0]},get content(){return(0,_polymerDom.dom)(this.$.mainSlot).getDistributedNodes()[0]},get scroller(){return this.$.mainContainer},get _headerMaxDelta(){return this.keepCondensedHeader?this._headerMargin:this.headerHeight},get _headerMargin(){return this.headerHeight-this.condensedHeaderHeight},_y:0,_prevScrollTop:0,measureHeaderHeight:function(){var header=this.header;if(header&&header.offsetHeight){this.headerHeight=header.offsetHeight}},scroll:function(top,smooth){if(smooth){var easingFn=function easeOutQuad(t,b,c,d){t/=d;return-c*t*(t-2)+b},animationId=Math.random(),duration=200,startTime=Date.now(),currentScrollTop=this.scroller.scrollTop,deltaScrollTop=top-currentScrollTop;this._currentAnimationId=animationId;(function updateFrame(){var now=Date.now(),elapsedTime=now-startTime;if(elapsedTime>duration){this.scroller.scrollTop=top;this._updateScrollState(top)}else if(this._currentAnimationId===animationId){this.scroller.scrollTop=easingFn(elapsedTime,currentScrollTop,deltaScrollTop,duration);requestAnimationFrame(updateFrame.bind(this))}}).call(this)}else{this.scroller.scrollTop=top;this._updateScrollState(top)}},condense:function(smooth){if(this.condenses&&!this.fixed&&!this.noReveal){switch(this.headerState){case 1:this.scroll(this.scroller.scrollTop-(this._headerMaxDelta-this._headerMargin),smooth);break;case 0:case 3:this.scroll(this._headerMargin,smooth);break;}}},scrollToTop:function(smooth){this.scroll(0,smooth)},_headerHeightChanged:function(headerHeight){if(null!==this._defaultCondsensedHeaderHeight){this._defaultCondsensedHeaderHeight=Math.round(1*headerHeight/3);this.condensedHeaderHeight=this._defaultCondsensedHeaderHeight}},_condensedHeaderHeightChanged:function(condensedHeaderHeight){if(condensedHeaderHeight){if(this._defaultCondsensedHeaderHeight!=condensedHeaderHeight){this._defaultCondsensedHeaderHeight=null}}},_condensesChanged:function(){this._updateScrollState(this.scroller.scrollTop);this._condenseHeader(null)},_setup:function(){var s=this.scroller.style;s.paddingTop=this.fixed?"":this.headerHeight+"px";s.top=this.fixed?this.headerHeight+"px":"";if(this.fixed){this._setHeaderState(0);this._transformHeader(null)}else{switch(this.headerState){case 1:this._transformHeader(this._headerMaxDelta);break;case 2:this._transformHeader(this._headerMargin);break;}}},_transformHeader:function(y){this._translateY(this.$.headerContainer,-y);if(this.condenses){this._condenseHeader(y)}this.fire("paper-header-transform",{y:y,height:this.headerHeight,condensedHeight:this.condensedHeaderHeight})},_condenseHeader:function(y){var reset=null===y;if(!this.scrollAwayTopbar&&this.header&&this.header.$&&this.header.$.topBar){this._translateY(this.header.$.topBar,reset?null:Math.min(y,this._headerMargin))}if(!this.noDissolve){this.$.headerBg.style.opacity=reset?"":(this._headerMargin-y)/this._headerMargin}this._translateY(this.$.headerBg,reset?null:y/2);if(!this.noDissolve){this.$.condensedHeaderBg.style.opacity=reset?"":y/this._headerMargin;this._translateY(this.$.condensedHeaderBg,reset?null:y/2)}},_translateY:function(node,y){this.transform(null===y?"":"translate3d(0, "+y+"px, 0)",node)},_scroll:function(event){if(this.header){this._updateScrollState(this.scroller.scrollTop);this.fire("content-scroll",{target:this.scroller},{cancelable:!1})}},_updateScrollState:function(scrollTop){var _Mathmax=Math.max,deltaScrollTop=scrollTop-this._prevScrollTop,y=_Mathmax(0,this.noReveal?scrollTop:this._y+deltaScrollTop);if(y>this._headerMaxDelta){y=this._headerMaxDelta;if(this.keepCondensedHeader){this._setHeaderState(2)}else{this._setHeaderState(1)}}else if(this.condenses&&scrollTop>=this._headerMargin){y=_Mathmax(y,this._headerMargin);this._setHeaderState(2)}else if(0===y){this._setHeaderState(0)}else{this._setHeaderState(3)}if(!this.fixed&&y!==this._y){this._transformHeader(y)}this._prevScrollTop=_Mathmax(scrollTop,0);this._y=y}});_exports.PaperScrollHeaderPanel=PaperScrollHeaderPanel;PaperScrollHeaderPanel.HEADER_STATE_EXPANDED=0;PaperScrollHeaderPanel.HEADER_STATE_HIDDEN=1;PaperScrollHeaderPanel.HEADER_STATE_CONDENSED=2;PaperScrollHeaderPanel.HEADER_STATE_INTERPOLATED=3});
